src/main/java/duke/tasks/TaskList.java

```diff
@@ -0,0 +1,173 @@
+package duke.tasks;
+
+import java.time.LocalDateTime;
+import java.util.ArrayList;
+import java.util.List;
+
+import duke.exceptions.InvalidDeadlineException;
+import duke.exceptions.InvalidEventException;
+import duke.exceptions.InvalidTodoDescriptionException;
+import duke.exceptions.UnknownMessageException;
+import duke.ui.Ui;
+
+/**
+ * The {@code TaskManager} class is responsible for managing a collection of tasks.
+ * It supports adding, retrieving, deleting, and identifying task types. Tasks
+ * include ToDos, Deadlines, and Events.
+ * <p>
+ * The class manages an internal list of {@link Task} objects and provides methods
+ * for interacting with and manipulating this list. It also handles errors related
+ * to invalid task inputs or operations.
+ * </p>
+ */
+public class TaskList {
+
+    /**
+     * The list of tasks being managed by the {@code TaskManager}.
+     */
+    private List<Task> tasks;
+
+    /**
+     * Constructs a new {@code TaskManager} with an empty task list.
+     */
+    public TaskList() {
+        this.tasks = new ArrayList<>();
+    }
+
+    /**
+     * Retrieves the list of tasks managed by this {@code TaskManager}.
+     *
+     * @return an {@code ArrayList} of tasks.
+     */
+    public List<Task> getTasks() {
+        return tasks;
+    }
+
+    /**
+     * Retrieves the task at the specified index in the task list.
+     *
+     * @param index the index of the task to retrieve.
+     * @return the {@code Task} at the specified index.
+     * @throws IndexOutOfBoundsException if the index is out of the range of the task list.
+     */
+    public Task getTask(int index) {
+        if (index >= 0 && index < tasks.size()) {
+            return tasks.get(index);
+        }
+        throw new IndexOutOfBoundsException("Invalid task index.");
+    }
+
+    /**
+     * Retrieves the last task in the task list.
+     *
+     * @return the last {@code Task} in the task list.
+     * @throws IllegalStateException if the task list is empty.
+     */
+    public Task getLastTask() {
+        if (tasks.isEmpty()) {
+            throw new IllegalStateException("No tasks available.");
+        }
+        return tasks.get(tasks.size() - 1);
+    }
+
+    public List<Task> getTasksOccurring(LocalDateTime dateTime) {
+        List<Task> tasksOccurringOnDateTime = new ArrayList<>();
+
+        for (Task task : this.tasks) {
+            if (task.occurring(dateTime)) {
+                tasksOccurringOnDateTime.add(task);
+            }
+        }
+
+        return tasksOccurringOnDateTime;
+    }
+
+    public void setTasks(List<Task> tasks) {
+        this.tasks = tasks;
+    }
+
+    /**
+     * Deletes the task at the specified index from the task list.
+     *
+     * @param index the index of the task to delete.
+     * @return the {@code Task} that was removed.
+     * @throws IndexOutOfBoundsException if the index is out of the range of the task list.
+     */
+    public Task deleteTask(int index) {
```

Comment:
I think this exception should be caught, otherwise, an invalid index specified by the user will crash your whole program